From da13632c7ab3f9e300c88b9117f8085ec7db207e Mon Sep 17 00:00:00 2001
From: Vincent Abriou <vincent.abriou@st.com>
Date: Fri, 21 Oct 2016 13:22:22 +0200
Subject: [PATCH 20/24] STM: waylandsink: memory:DMABuf preferred while
 negotiation

Change the ranking of memory:DMABuf in order to negotiate DMABUF as
preferred buffer type.

Change-Id: Ifc5a9d41228c44345608e34dfea55dd10297ba1a
Signed-off-by: Vincent Abriou <vincent.abriou@st.com>
Reviewed-on: https://gerrit.st.com/59311
Reviewed-by: Jean Christophe TROTIN <jean-christophe.trotin@st.com>
Tested-by: Jean Christophe TROTIN <jean-christophe.trotin@st.com>
---
 ext/wayland/gstwaylandsink.c | 26 +++++++++++++++++---------
 1 file changed, 17 insertions(+), 9 deletions(-)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index acd7c90..2b51b20 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -78,9 +78,9 @@ GST_DEBUG_CATEGORY (gstwayland_debug);
 static GstStaticPadTemplate sink_template = GST_STATIC_PAD_TEMPLATE ("sink",
     GST_PAD_SINK,
     GST_PAD_ALWAYS,
-    GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE (WL_VIDEO_FORMATS) ";"
-        GST_VIDEO_CAPS_MAKE_WITH_FEATURES (GST_CAPS_FEATURE_MEMORY_DMABUF,
-            WL_VIDEO_FORMATS))
+    GST_STATIC_CAPS (GST_VIDEO_CAPS_MAKE_WITH_FEATURES
+        (GST_CAPS_FEATURE_MEMORY_DMABUF,
+            WL_VIDEO_FORMATS) ";" GST_VIDEO_CAPS_MAKE (WL_VIDEO_FORMATS))
     );
 
 static void gst_wayland_sink_get_property (GObject * object,
@@ -441,7 +441,7 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
     GValue shm_list = G_VALUE_INIT, dmabuf_list = G_VALUE_INIT;
     GValue value = G_VALUE_INIT;
     GArray *formats;
-    gint i;
+    gint i, num_structures;
     uint fmt;
 
     g_value_init (&shm_list, GST_TYPE_LIST);
@@ -456,9 +456,6 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
       gst_value_list_append_value (&shm_list, &value);
     }
 
-    gst_structure_set_value (gst_caps_get_structure (caps, 0), "format",
-        &shm_list);
-
     /* Add corresponding dmabuf formats */
     formats = sink->display->dmabuf_formats;
     for (i = 0; i < formats->len; i++) {
@@ -466,8 +463,19 @@ gst_wayland_sink_get_caps (GstBaseSink * bsink, GstCaps * filter)
       g_value_set_string (&value, gst_wl_dmabuf_format_to_string (fmt));
       gst_value_list_append_value (&dmabuf_list, &value);
     }
-    gst_structure_set_value (gst_caps_get_structure (caps, 1), "format",
-        &dmabuf_list);
+
+    /* Set supported format in the right structure */
+    num_structures = gst_caps_get_size (caps);
+    for (i = 0; i < num_structures; i++) {
+      if (gst_caps_features_contains (gst_caps_get_features (caps, i),
+              GST_CAPS_FEATURE_MEMORY_SYSTEM_MEMORY))
+        gst_structure_set_value (gst_caps_get_structure (caps, i), "format",
+            &shm_list);
+      else if (gst_caps_features_contains (gst_caps_get_features (caps, i),
+              GST_CAPS_FEATURE_MEMORY_DMABUF))
+        gst_structure_set_value (gst_caps_get_structure (caps, i), "format",
+            &dmabuf_list);
+    }
 
     g_value_unset (&value);
     g_value_unset (&dmabuf_list);
-- 
2.7.4

